File: server/app/services/recipes_service.py
from typing import List

from app.extensions import db
from app.models.recipe import Recipe
from app.models.recipe_ingredient import RecipeIngredient


def get_all_recipes(user_id: int) -> List[Recipe]:
    return (
        Recipe.query.filter(Recipe.user_id == user_id)
        .order_by(Recipe.created_at.desc())
        .all()
    )


def get_recipe_by_id(recipe_id: int, user_id: int) -> Recipe | None:
    return Recipe.query.filter(
        Recipe.id == recipe_id,
        Recipe.user_id == user_id,
    ).first()


def create_recipe(user_id: int, payload: dict) -> Recipe:
    title = payload.get("title")
    if not title:
        raise ValueError("Recipe title is required")

    servings = payload.get("servings", 1)
    ingredients = payload.get("ingredients", [])

    recipe = Recipe(
        user_id=user_id,
        title=title.strip(),
        description=payload.get("description"),
        servings=servings,
    )

    for idx, ing in enumerate(ingredients):
        recipe.ingredients.append(
            RecipeIngredient(
                name=ing["name"].strip(),
                quantity=ing.get("quantity"),
                unit=ing.get("unit"),
                notes=ing.get("notes"),
                sort_order=idx,
            )
        )

    db.session.add(recipe)
    db.session.commit()
    return recipe


def update_recipe(recipe: Recipe, payload: dict) -> Recipe:
    if "title" in payload:
        recipe.title = payload["title"].strip()

    if "description" in payload:
        recipe.description = payload["description"]

    if "servings" in payload:
        recipe.servings = payload["servings"]

    if "ingredients" in payload:
        recipe.ingredients.clear()
        for idx, ing in enumerate(payload["ingredients"]):
            recipe.ingredients.append(
                RecipeIngredient(
                    name=ing["name"].strip(),
                    quantity=ing.get("quantity"),
                    unit=ing.get("unit"),
                    notes=ing.get("notes"),
                    sort_order=idx,
                )
            )

    db.session.commit()
    return recipe


def delete_recipe(recipe: Recipe) -> None:
    db.session.delete(recipe)
    db.session.commit()

File: client/src/services/recipes.service.js
import { http } from "./http.js";

/**
 * @typedef {Object} RecipeIngredient
 * @property {number} [id]
 * @property {number} [recipeId]
 * @property {string} name
 * @property {number|null} [quantity]
 * @property {string|null} [unit]
 * @property {string|null} [notes]
 * @property {number} [sortOrder]
 */

/**
 * @typedef {Object} Recipe
 * @property {number} id
 * @property {number} userId
 * @property {string} title
 * @property {string|null} description
 * @property {number} servings
 * @property {number|null} [prepMin]
 * @property {number|null} [cookMin]
 * @property {string|null} [instructions]
 * @property {string|null} [notes]
 * @property {string} createdAt
 * @property {string} updatedAt
 * @property {RecipeIngredient[]} [ingredients]
 */

/**
 * @typedef {Object} RecipeCreatePayload
 * @property {string} title
 * @property {string|null} [description]
 * @property {number} [servings]
 * @property {number|null} [prepMin]
 * @property {number|null} [cookMin]
 * @property {string|null} [instructions]
 * @property {string|null} [notes]
 * @property {RecipeIngredient[]} [ingredients]
 */

/**
 * @typedef {Object} RecipeUpdatePayload
 * @property {string} [title]
 * @property {string|null} [description]
 * @property {number} [servings]
 * @property {number|null} [prepMin]
 * @property {number|null} [cookMin]
 * @property {string|null} [instructions]
 * @property {string|null} [notes]
 * @property {RecipeIngredient[]} [ingredients]
 */

export const recipesService = {
    /**
     * @returns {Promise<Recipe[]>}
     */
    list() {
        return http.get("/recipes");
    },

    /**
     * @param {number} recipeId
     * @returns {Promise<Recipe>}
     */
    get(recipeId) {
        return http.get(`/recipes/${recipeId}`);
    },

    /**
     * @param {RecipeCreatePayload} payload
     * @returns {Promise<Recipe>}
     */
    create(payload) {
        return http.post("/recipes", payload);
    },

    /**
     * @param {number} recipeId
     * @param {RecipeUpdatePayload} payload
     * @returns {Promise<Recipe>}
     */
    update(recipeId, payload) {
        return http.patch(`/recipes/${recipeId}`, payload);
    },

    /**
     * @param {number} recipeId
     * @returns {Promise<{status: string}>}
     */
    remove(recipeId) {
        return http.delete(`/recipes/${recipeId}`);
    },

    publish(recipeId) {
        return http.post(`/recipes/${recipeId}/publish`, {});
    },
    unpublish(recipeId) {
        return http.post(`/recipes/${recipeId}/unpublish`, {});
    },

};
